Specify table via query in tests and various other changes related to metadata and tests

bertfrees opened the pull request:
- Specify table via query in tests: see https://github.com/liblouis/liblouis/issues/413
- Fix hyphenation tests
- Support "." in metadata: fixes https://github.com/liblouis/liblouis/issues/359
- Various changes to table metadata

egli commented on tests/check_included_tables.sh:
```diff
@@ -18,14 +18,17 @@ has_metadata() {
-done
+done <<< $(
+    for table in $(dirname $0)/../tables/*; do
+        cat "$table" | dos2unix | grep '^include ' | sed "s|^include  *\([^ ]*\).*$|$(dirname $table)/\1|g"
```
Not everybody has `dos2unix` installed. Do we really need it?

bertfrees replied:
For some reason the `sed` acted weird for some tables which turned out to be DOS encoded. Because I didn't know how to fix it I added the `dos2unix`. Anyway, since I disabled this test, and the file is only kept for the purpose of possible enabling it again later, I don't think it really matters that much right now. Maybe just add a small comment?

egli replied:
I guess we should go through the tables and make sure they're all utf8 just like @mhameed did once. That would be better.

bertfrees replied:
They are UTF8, some just have Windows line endings.

egli commented on tables/fr-bfu-comp6.utb:
```diff
@@ -1,5 +1,21 @@
+#+locale:fr
+#+type:computer
+#+dots:6
+#+bfu
```
What is `bfu`?

bertfrees replied:
Braille Français Unifié. It's just a way to differentiate with other French tables. If we think this table should be the default for French computer braille, then there is a problem with the metadata.

egli replied:
But shouldn't this be some form of metadata just like ueb? Ah I see that ueb is handled just like that. But then shouldn't meta data always be `key: value`?

bertfrees replied:
Both are supported.

bertfrees replied:
See http://liblouis.org/documentation/liblouis.html#Table-Metadata

egli replied:
Oh, I didn't know that. I'd only support `key: value`. a) to be orthogonal and b) to make the syntax easier on the user

bertfrees replied:
You are a bit late with that feedback :) I liked it. But can we leave that discussion for another time?

egli commented on tables/fr-fr-g1.utb:
```diff
@@ -1,5 +1,22 @@
+#+type:literary
+#+contraction:no
+#+grade:1
+#+bfu
```
Again `bfu`?

egli commented on tables/fr.tbl:
```diff
@@ -2,6 +2,7 @@
 #+type:literary
 #+contraction:full
 #+grade:2
+#+bfu
```
`bfu`?

bertfrees commented:
Thanks for the feedback!

> 1. Too many changes in one PR. IMHO they should be in separate PRs
>    - split the yaml tests
>    - support for finding tables based on meta data
>    - some hyphenation extension

Ideally yes (although I think [my division](https://github.com/liblouis/liblouis/pull/437#issue-271128692) was better). But is it really worth the trouble? I already split up everything nicely in commits which is the main thing (some people wouldn't even bother with that). Splitting it up in branches and PRs can slightly improve the readability of the git tree and the issue history on Github, but I felt lazy. You just gain so little with it.

> 2. While the meta data based YAML tests are nice they also add an indirection which makes it hard to find out which table is actually tested now

I think we have to get used to identifying tables in terms of metadata instead of file names. That's the whole purpose of this. I have hesitated a bit but in the end decided to leave the `# table: [...]`comments in the tests, just as a transitional measure. The idea is to remove them eventually. Would you feel better if we ensure somehow in the test that the table actually used matches an expected value?

> 3. The split of the yaml test suite. I'd prefer if we could find a way to have only one test run. Maybe we could define an additional log compiler that would handle the braille_spec tests slightly differently from the other yaml tests.

I first thought about creating a special environment variable that lou_checkyaml would use to index the tables. But that just seemed wrong.

The real problem is that different tests in the same directory can not have a different set of environment variables. Unless you set the environment variables in the log compiler. Maybe that is what you are suggesting? But can this be done without changing the file extensions?

egli commented:
>> 1. Too many changes in one PR. IMHO they should be in separate PRs 

> Ideally yes (although I think my division was better). But is it really worth the trouble?

It's just that some things are no-brainers. They could go in. Others maybe need more discussion. But I guess we could cherry pick, so it's not such a big deal.

>> While the meta data based YAML tests are nice they also add an indirection which makes it hard to find out which table is actually tested now

> I think we have to get used to identifying tables in terms of metadata instead of file names. That's the whole purpose of this. I have hesitated a bit but in the end decided to leave the # table: [...]comments in the tests, just as a transitional measure. The idea is to remove them eventually. Would you feel better if we ensure somehow in the test that the table actually used matches an expected value?

The `# table: [...]` comment is what made me stumble at the beginning. It feels like it should be removed. I don't know what would make me feel better :-), but I guess making sure there is an actual matching table is a first step

> The real problem is that different tests in the same directory can not have a different set of environment variables. Unless you set the environment variables in the log compiler. Maybe that is what you are suggesting? But can this be done without changing the file extensions?

I had something like that in mind, but I'm not sure if it goes far enough. Maybe it really is a simple match between extension and "compiler". Yes, that is how it is, I just [read up on it](https://www.gnu.org/software/automake/manual/html_node/Parallel-Test-Harness.html#Parallel-Test-Harness).

One option that I could see is to create a wrapper script that invokes the tests with a different environment depending on whether they are braille_spec tests or not. Not sure if that really simplifies things

bertfrees commented:
> making sure there is an actual matching table is a first step

What should that look like in YAML?

> One option that I could see is to create a wrapper script that invokes the tests with a different environment depending on whether they are braille_spec tests or not.

Yeah that would work. I'll try it.

egli commented:
>> making sure there is an actual matching table is a first step

> What should that look like in YAML?

I thought you could just error out if you didn't find exactly one matching table for the query. Or you could just error out if you didn't find at least one matching table.

bertfrees commented:
With the current query API there can be at most one match. If there is no match, the test errors out.

Actually I'm thinking about extending this idea to multiple matches. Then you can for example create a set of tests that should be satisfied for all German tables regardless of grade. You can do that already but you need to list all the combinations explicitly.

Example:
```
table:
  locale: de
  grade: 2
table:
  locale: de
  grade: 1
table:
  locale: de
  grade: 0
```

By supporting multiple matches this would become:
```
table:
  locale: de
```

bertfrees commented:
@egli I've combined all the tests into a single run again.

bertfrees commented:
@egli I've combined all the tests into a single run again.

bertfrees commented:
I have some more metadata changes coming up so maybe we can wait for that before merging. But are the current changes approved now?

egli commented:
Oh, I had in my notes that this could go in but you're telling me to wait, no problem.

> But are the current changes approved now?

I don't like the metadata that is not a key-value tuple, but aside from that I think I'm fine with this change

bertfrees commented:
OK thanks.

I don't think we need to solve the key-value now. It is currently allowed so why not use it. The idea for it was to support attaching a list of tags to a table, which can be used for discovery. Compare it with [`keywords` metadata in HTML](https://www.w3.org/TR/html/document-metadata.html#keywords) used for search engine optimization, keywords in a paper abstract, etc.

If you'd like to remove this feature or implement it differently we can discuss it of course.

Anyway, because we don't have a standardized field for metadata like "ueb" and "bfu" I decided to just use simple tags for it for now. I have also been thinking about creating an official field named "code". The meaning would be: a general identifier of the braille code, which does not identify variants such as grades.

bertfrees commented:
@egli
- I've removed the metadata fields without a value, that you didn't like. They have been replaced with e.g. `#+system: ueb`.
- I added a (temporary) feature to assert that the table that is actually tested is your expected table. The `# table: [...]` comments are gone. Still eventually we should move to true metadata based tests, but the check is nice to have for now. It works like this:
   ```yaml
   table:
     locale: de
     __match: de-de-g2.ctb
   ```
- And I did some other changes to metadata.

I did a forced push.

I have two more commits for which I will make a new PR when this one is merged. One adds `#-name` fields and the other adds `#-display-name` fields. The `display-name` fields are generated automatically based on metadata with a script I wrote.

bertfrees commented:
To do:
- [ ] documentation

egli commented on tools/lou_checkyaml.c:
```diff
@@ -127,11 +166,13 @@ read_table(yaml_event_t *start_event, yaml_parser_t *parser) {
-				"Expected %s or %s (actual %s)", event_names[YAML_SEQUENCE_START_EVENT],
-				event_names[YAML_SCALAR_EVENT], event_names[event.type]);
-
+				"Expected %s or %s or %s (actual %s)",
```
Nit pick: This should be "Expected %s, %s or %s"

egli commented on tools/lou_checkyaml.c:
```diff
@@ -180,6 +221,32 @@ read_table(yaml_event_t *start_event, yaml_parser_t *parser) {
+		query = read_table_query(parser, &table_file_name_check);
+		table = lou_findTable(query);
+		free(query);
+		if (!table) exit(EXIT_FAILURE);
```
Shouldn't we print an error message here before dying?

bertfrees replied:
Yes you're right.

egli commented on tools/lou_checkyaml.c:
```diff
@@ -668,20 +735,34 @@ main(int argc, char *argv[]) {
+	int has_next;
+	has_next = yaml_parser_parse(&parser, &event);
+
+	const char *display_table = NULL;
```
I don't quite understand this. What is the YAML syntax here? Where do you define a display table? And why do you need this? Can't you just define the display table with the table? Ah, maybe you can't when doing a table query. Is that it?

Ah, never mind I can see it in the YAML files.

bertfrees replied:
Yes, that's it. And a second reason is that I want to have display and translation tables separated in the code (https://github.com/liblouis/liblouis/issues/307). Separating them in the tests is already a first step.

bertfrees commented:
Two commits were already cherry-picked so I'm cutting them out of this branch: f6b19b35f3819d9e8af611d952876157063f0ea2 and 43bdfc14d0fddda4e5036ac7ce6cebf288fba2b2.

bertfrees commented:
Merged latest master. @egli if you can approve I push to master.

egli commented on doc/liblouis.texi:
```diff
@@ -2465,26 +2465,56 @@ tests:
 @end example
 
-The three basic components of a test file are as follows:
+The four basic components of a test file are as follows:
```
I don't quite understand why there are four components now. AFAICT you did not add a component to the test file. If I understand correctly the `table:` component can now also handle a table query, but then there would still be three components. Ah I guess it's the `display`. OK, I get it now

bertfrees replied:
Yeah.
